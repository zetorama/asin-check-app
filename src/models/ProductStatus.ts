export enum ProductStatus {
    New = 'new',
    InQueue = 'inqueue',
    Refreshing = 'refreshing',
    Success = 'success',
    Nomatch = 'nomatch',
    Timeout = 'timeout',
    Fail = 'fail',
}

export default ProductStatus
