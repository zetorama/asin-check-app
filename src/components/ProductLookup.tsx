import React, { Component } from 'react'
import { InputGroup, InputGroupAddon, InputGroupText, Input, Button, Form } from 'reactstrap'
import autobind from '../utils/autobind'

import { ProductAsin } from '../models/Product'

export interface ProductLookupProps {
    defaultValue?: string
    onLookup: (e: React.MouseEvent | React.FormEvent, value: string) => void
}

export interface ProductLookupState {
    defaultValue?: string
    value: string
}

let formId = 0

export class ProductLookup extends Component<ProductLookupProps, ProductLookupState> {
    static getDerivedStateFromProps(props: ProductLookupProps, state: ProductLookupState): ProductLookupState | null {
        if (props.defaultValue !== state.defaultValue) {
            return {
                value: props.defaultValue || '',
                defaultValue: props.defaultValue,
            }
        }

        return null
    }

    private formId = ++formId

    state: ProductLookupState = {
        value: '',
    }

    render() {
        const { value, defaultValue } = this.state
        const isBtnDisabled = !value || value === defaultValue

        const searchId = `product-lookup-search-${this.formId}`

        return (
            <Form onSubmit={this.handleLookup}>
                <InputGroup className='mb-3'>
                    <InputGroupAddon addonType='prepend'>
                        <InputGroupText tag='label' htmlFor={searchId}>
                            ASIN
                        </InputGroupText>
                    </InputGroupAddon>
                    <Input
                        id={searchId}
                        pattern='(\W*\w{10,}\W*){1,}'
                        placeholder='ABCDEFG012'
                        className='text-monospace'
                        value={value}
                        onChange={this.handleInputChange}
                    />
                    <InputGroupAddon addonType='append'>
                        <Button type='submit' color='primary' disabled={isBtnDisabled}>
                            Find
                        </Button>
                    </InputGroupAddon>
                </InputGroup>
            </Form>
        )
    }

    @autobind
    handleInputChange(e: React.ChangeEvent<HTMLInputElement>) {
        this.setState({
            value: e.target.value,
        })
    }

    @autobind
    handleLookup(e: React.MouseEvent | React.FormEvent) {
        e.preventDefault()
        const { onLookup } = this.props
        onLookup && onLookup(e, this.state.value)
    }
}

export default ProductLookup
