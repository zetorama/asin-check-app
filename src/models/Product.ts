import { getParent, types, Instance, SnapshotIn } from 'mobx-state-tree'

import { StoreModel } from './Store'

export enum ProductStatus {
    Init = 'init',
    Queue = 'queue',
    Fetch = 'fetch',
    Success = 'success',
    Nomatch = 'nomatch',
    Timeout = 'timeout',
    Fail = 'fail',
}

export const ProductDetails = types.model('ProductDetails', {
    title: types.string,
    categories: types.array(types.string),
    dimensions: types.maybe(types.string),
    rank: types.maybe(types.array(types.string)),
    rating: types.maybe(types.string),
})

export type ProductAsin = string

export type ProductModel = Instance<typeof Product>
export type ProductSnapshot = SnapshotIn<typeof Product>

export const Product = types
    .model('Product', {
        asin: types.identifier,
        status: types.enumeration<ProductStatus>('ProductStatus', Object.values(ProductStatus)),
        data: types.maybe(types.frozen(ProductDetails)),

        createdOn: types.maybe(types.string),
        updatedOn: types.maybe(types.string),
        isViewed: false,
    })
    .views((self) => ({
        get isLoading(): boolean {
            return (
                self.status === ProductStatus.Init ||
                self.status === ProductStatus.Queue ||
                self.status === ProductStatus.Fetch
            )
        },
        get isCurrentlyViewing(): boolean {
            // assumming, parent has an `activeProduct` as a ref
            return getParent<StoreModel>(self, 2).viewingProduct === self
        },
    }))
    .actions((self) => ({
        toggleViewed(setAsViewed: boolean = !self.isViewed) {
            self.isViewed = setAsViewed
        },
    }))

export default Product
