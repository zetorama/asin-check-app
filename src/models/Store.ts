import { types, getParent, getSnapshot, applySnapshot, Instance, SnapshotIn } from 'mobx-state-tree'

import { fetchProductsLog, saveProductsLog, fetchProductsByAsins, refreshProductsByAsins } from '../services/product'
import { Product, ProductAsin, ProductModel, ProductStatus, ProductSnapshot } from './Product'

export const PRODUCTS_LIMIT = 12

export type StoreModel = Instance<typeof Store>
export type StoreSnapshot = SnapshotIn<typeof Store>

type Void<T> = { [P in keyof T]?: undefined }

const resetObj = <T>(source: T): Void<T> =>
    Object.keys(source).reduce((all, one) => Object.assign(all, { [one]: undefined }), {})

export const ProductInStore = types.compose(
    Product,
    types.model('ProductInStore').views((self) => ({
        get isCurrentlyViewing(): boolean {
            // assumming, parent has an `activeProduct` as a ref
            return getParent<StoreModel>(self, 2).viewingProduct === self
        },
    })),
)

export type ProductInStoreModel = Instance<typeof ProductInStore>
export type ProductInStoreSnapshot = SnapshotIn<typeof ProductInStore>

export const Store = types
    .model('AppStore', {
        version: '0.0.0',
        products: types.array(ProductInStore),
        viewingProduct: types.maybe(types.reference(ProductInStore)),
        lookingUpValue: types.optional(types.string, ''),
    })
    .views((self) => ({
        get loadingProducts(): ProductModel[] {
            return self.products.filter((p) => p.isLoading)
        },
        getProduct(asin: ProductAsin): ProductModel | undefined {
            return self.products.find((p) => p.asin === asin)
        },
    }))
    .actions((self) => ({
        mergeUpdates(products: ProductSnapshot[]) {
            // Update only existed items, keeping the order
            for (const update of products) {
                const product = self.getProduct(update.asin)
                if (product) {
                    applySnapshot(product, Object.assign(resetObj(product), update))

                    if (product === self.viewingProduct) {
                        // Mark as seen, as its currently opened
                        product.toggleViewed(true)
                    }
                }
            }
        },
        createProduct(asin: ProductAsin): ProductInStoreModel {
            const product = self.getProduct(asin)
            const defaults = product ? getSnapshot(product) : undefined
            return ProductInStore.create({
                ...defaults,
                asin,
                status: ProductStatus.New,
            })
        },
        addProducts(products: ProductModel[]) {
            if (!products || !Array.isArray(products) || !products.length) return

            const oldProducts = self.products.filter((p1) => !products.find((p2) => p1.asin === p2.asin))
            const newProducts: ProductInStoreModel[] = products.concat(oldProducts).slice(0, PRODUCTS_LIMIT)
            self.products.replace(newProducts)
        },
        clearLookingUpValue() {
            self.lookingUpValue = ''
        },
        viewProduct(product?: ProductModel | ProductAsin) {
            self.viewingProduct = product
                ? self.getProduct(ProductInStore.is(product) ? product.asin : product)
                : undefined
            if (self.viewingProduct && !self.viewingProduct.isViewed) {
                self.viewingProduct.toggleViewed(true)
            }
        },
    }))
    .actions((self) => ({
        saveProductsLog() {
            saveProductsLog(getSnapshot(self.products))
        },
        async fetchProductsLog() {
            const products = (await fetchProductsLog()) as ProductModel[]
            products && products.length && self.addProducts(products)
        },
        async lookup(value: string) {
            self.lookingUpValue = String(value || '')

            if (self.lookingUpValue) {
                const asins = self.lookingUpValue.split(/\W+/)
                const products = asins
                    .filter((asin, index) => asin && asins.indexOf(asin) === index)
                    .map((asin) => self.createProduct(asin))

                self.viewingProduct = undefined
                await self.addProducts(products)

                // switch to a newly added product
                self.viewProduct(self.products[0])
                self.clearLookingUpValue()
            }
        },
        async fetchUpdates(asins: ProductAsin[] = self.loadingProducts.map((p) => p.asin)) {
            const update = await fetchProductsByAsins(asins)
            update && self.mergeUpdates(update)
        },
        async refreshProducts(asins: ProductAsin[]) {
            for (const asin of asins) {
                const product = self.getProduct(asin)
                product && product.setStatus(ProductStatus.InQueue)
            }
            const update = await refreshProductsByAsins(asins)
            update && self.mergeUpdates(update)
        },
    }))

export default Store
