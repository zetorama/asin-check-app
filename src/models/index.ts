import { Store, StoreModel, StoreSnapshot } from './Store'

export * from './Store'
export * from './Product'

export function createStore(initial?: StoreSnapshot): StoreModel {
    return Store.create(initial)
}

export default createStore
