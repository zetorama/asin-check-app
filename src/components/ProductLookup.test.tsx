import * as React from 'react'
import { shallow, mount } from 'enzyme'

import { ProductLookup } from './ProductLookup'
import { Product, ProductStatus } from '../models'

describe('ProductLookup', () => {
    it('renders with empty value', () => {
        const cb = jest.fn()
        const cmp = shallow(<ProductLookup onLookup={cb} />)

        expect(
            cmp
                .find('.lookup-button')
                .at(0)
                .prop('disabled'),
        ).toBe(true)

        expect(cmp).toMatchSnapshot()
    })

    it('renders with default value', () => {
        const cb = jest.fn()
        const cmp = shallow(<ProductLookup onLookup={cb} defaultValue={'foo bar'} />)

        expect(
            cmp
                .find('.lookup-button')
                .at(0)
                .prop('disabled'),
        ).toBe(true)

        expect(cmp).toMatchSnapshot()
    })

    it('changes input value', () => {
        const cb = jest.fn()
        const cmp = mount(<ProductLookup onLookup={cb} defaultValue={'foo bar'} />)

        expect(
            cmp
                .find('.lookup-input')
                .at(0)
                .prop('value'),
        ).toBe('foo bar')
        cmp.find('.lookup-input')
            .at(0)
            .simulate('change', { target: { value: 'xxx' } })
        expect(
            cmp
                .find('.lookup-input')
                .at(0)
                .prop('value'),
        ).toBe('xxx')

        expect(
            cmp
                .find('.lookup-button')
                .at(0)
                .prop('disabled'),
        ).toBe(false)

        expect(cmp).toMatchSnapshot()
    })

    it.skip('calls lookup callback', () => {
        const cb = jest.fn()
        const cmp = mount(<ProductLookup onLookup={cb} />)

        cmp.find('.lookup-input')
            .at(0)
            .simulate('change', { target: { value: 'xxx' } })

        expect(
            cmp
                .find('.lookup-button')
                .at(0)
                .prop('disabled'),
        ).toBe(false)
        cmp.find('.lookup-button')
            .at(0)
            .simulate('click')

        // TODO: figure out, why click doesn't workâ€¦
        expect(cb).toBeCalled()
    })
})
