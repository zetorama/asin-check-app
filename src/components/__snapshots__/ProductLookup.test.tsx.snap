// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductLookup changes input value 1`] = `
<ProductLookup
  defaultValue="foo bar"
  onLookup={[MockFunction]}
>
  <Form
    onSubmit={[Function]}
    tag="form"
  >
    <form
      className=""
      onSubmit={[Function]}
    >
      <InputGroup
        className="mb-3"
        tag="div"
      >
        <div
          className="mb-3 input-group"
        >
          <InputGroupAddon
            addonType="prepend"
            tag="div"
          >
            <div
              className="input-group-prepend"
            >
              <InputGroupText
                htmlFor="product-lookup-search-3"
                tag="label"
              >
                <label
                  className="input-group-text"
                  htmlFor="product-lookup-search-3"
                >
                  ASIN
                </label>
              </InputGroupText>
            </div>
          </InputGroupAddon>
          <Input
            className="lookup-input text-monospace"
            id="product-lookup-search-3"
            onChange={[Function]}
            pattern="(\\\\W*\\\\w{10,}\\\\W*){1,}"
            placeholder="ABCDEFG012"
            type="text"
            value="xxx"
          >
            <input
              className="lookup-input text-monospace form-control"
              id="product-lookup-search-3"
              onChange={[Function]}
              pattern="(\\\\W*\\\\w{10,}\\\\W*){1,}"
              placeholder="ABCDEFG012"
              type="text"
              value="xxx"
            />
          </Input>
          <InputGroupAddon
            addonType="append"
            tag="div"
          >
            <div
              className="input-group-append"
            >
              <Button
                className="lookup-button"
                color="primary"
                disabled={false}
                tag="button"
                type="submit"
              >
                <button
                  aria-label={null}
                  className="lookup-button btn btn-primary"
                  disabled={false}
                  onClick={[Function]}
                  type="submit"
                >
                  Find
                </button>
              </Button>
            </div>
          </InputGroupAddon>
        </div>
      </InputGroup>
    </form>
  </Form>
</ProductLookup>
`;

exports[`ProductLookup renders with default value 1`] = `
<Form
  onSubmit={[Function]}
  tag="form"
>
  <InputGroup
    className="mb-3"
    tag="div"
  >
    <InputGroupAddon
      addonType="prepend"
      tag="div"
    >
      <InputGroupText
        htmlFor="product-lookup-search-2"
        tag="label"
      >
        ASIN
      </InputGroupText>
    </InputGroupAddon>
    <Input
      className="lookup-input text-monospace"
      id="product-lookup-search-2"
      onChange={[Function]}
      pattern="(\\\\W*\\\\w{10,}\\\\W*){1,}"
      placeholder="ABCDEFG012"
      type="text"
      value="foo bar"
    />
    <InputGroupAddon
      addonType="append"
      tag="div"
    >
      <Button
        className="lookup-button"
        color="primary"
        disabled={true}
        tag="button"
        type="submit"
      >
        Find
      </Button>
    </InputGroupAddon>
  </InputGroup>
</Form>
`;

exports[`ProductLookup renders with empty value 1`] = `
<Form
  onSubmit={[Function]}
  tag="form"
>
  <InputGroup
    className="mb-3"
    tag="div"
  >
    <InputGroupAddon
      addonType="prepend"
      tag="div"
    >
      <InputGroupText
        htmlFor="product-lookup-search-1"
        tag="label"
      >
        ASIN
      </InputGroupText>
    </InputGroupAddon>
    <Input
      className="lookup-input text-monospace"
      id="product-lookup-search-1"
      onChange={[Function]}
      pattern="(\\\\W*\\\\w{10,}\\\\W*){1,}"
      placeholder="ABCDEFG012"
      type="text"
      value=""
    />
    <InputGroupAddon
      addonType="append"
      tag="div"
    >
      <Button
        className="lookup-button"
        color="primary"
        disabled={true}
        tag="button"
        type="submit"
      >
        Find
      </Button>
    </InputGroupAddon>
  </InputGroup>
</Form>
`;
